[popover] Implement popovertarget & popovertargetaction attributes

nt1m opened the pull request:
#### 25998fb94754e1acb8d93fa0f51d0c1e1e694ba4
<pre>
[popover] Implement popovertarget &amp; popovertargetaction attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=252078">https://bugs.webkit.org/show_bug.cgi?id=252078</a>
rdar://105612966

Reviewed by Chris Dumez.

This adds support for the popoverTargetElement &amp; popoverTargetAction IDL attributes, which map to popovertarget &amp; popovertargetaction HTML attributes.

Spec change is at <a href="https://github.com/whatwg/html/pull/8962">https://github.com/whatwg/html/pull/8962</a> , WPT have already been updated by a Chromium commit to match.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-invoking-attribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-stacking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-combinations-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
Rebaseline tests with progressions. Remaining failures are either due to these unimplemented parts:
- invoker relationship removal handling (not standardized?)
- popover focusing steps (<a href="https://bugs.webkit.org/show_bug.cgi?id=253231)">https://bugs.webkit.org/show_bug.cgi?id=253231)</a>
- anchor attribute (not standardized)

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
Update list of built files.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
Check for invokers as described in algorithm in: <a href="https://html.spec.whatwg.org/#topmost-clicked-popover">https://html.spec.whatwg.org/#topmost-clicked-popover</a>

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::addRelations):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute):
(WebCore::Element::isElementsArrayReflectionAttribute):
(WebCore::Element::attributeChanged):
(WebCore::Element::getElementAttribute const):
(WebCore::Element::setElementAttribute):
(WebCore::Element::getElementsArrayAttribute const):
(WebCore::Element::setElementsArrayAttribute):
* Source/WebCore/dom/Element.h:
Add IDL element reflection support for popovertarget attribute. Simplified settings checks.

* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::invoker const):
(WebCore::PopoverData::setInvoker):
Store invoker in popover data.

* Source/WebCore/html/HTMLAttributeNames.in: Define popovertargetAttr &amp; popovertargetActionAttr.
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler): Invoke popover activation behavior.
* Source/WebCore/html/HTMLButtonElement.idl: Use PopoverInvokerElement mixin.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::topmostPopoverAncestor): Check for invokers as defined in the spec: <a href="https://html.spec.whatwg.org/#topmost-popover-ancestor">https://html.spec.whatwg.org/#topmost-popover-ancestor</a>
(WebCore::HTMLElement::hidePopoverInternal): Clear invoker as defined in the spec.
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::toggleAtom):
(WebCore::showAtom):
(WebCore::hideAtom):
(WebCore::HTMLFormControlElement::popoverTargetElement const): Implement <a href="https://html.spec.whatwg.org/#popover-target-element">https://html.spec.whatwg.org/#popover-target-element</a>
(WebCore::HTMLFormControlElement::popoverTargetAction const): Implement attribute reflection.
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const): Implement activation behavior: <a href="https://html.spec.whatwg.org/#popover-target-attribute-activation-behavior">https://html.spec.whatwg.org/#popover-target-attribute-activation-behavior</a>
* Source/WebCore/html/HTMLFormControlElement.h:
(WebCore::HTMLFormControlElement::setPopoverTargetAction): Implement attribute reflection.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler): Invoke popover activation behavior.
* Source/WebCore/html/HTMLInputElement.idl: Use PopoverInvokerElement mixin.
* Source/WebCore/html/PopoverInvokerElement.idl: Added.

Canonical link: <a href="https://commits.webkit.org/261346@main">https://commits.webkit.org/261346@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af4f69208c3617dd61d8691c24fd9a4f39e9ab1f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20526 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/3202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115343 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21893 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11620 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/3202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117153 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/44904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13027 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15497 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4309 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8992,15 +8993,36 @@ void Document::handlePopoverLightDismiss(PointerEvent& event)
+                    return targetPopover;
+            }
+            return nullptr;
+        }();
```
Looks like we're iterating over the same elements twice, can we rewrite to do this in a single loop?

nt1m replied:
yeah it's probably possible

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8992,15 +8993,36 @@ void Document::handlePopoverLightDismiss(PointerEvent& event)
+            if (!second)
+                return first;
+            Element* highestFound = nullptr;
+            for (auto& element : m_topLayerElements) {
```
Wouldn't we want to iterate backwards so we don't have to traverse the whole ListHashSet?

nt1m replied:
ah yes, very true

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2093,7 +2093,7 @@ void Element::attributeChanged(const QualifiedName& name, const AtomString& oldV
             }
         } else if (name == HTMLNames::partAttr)
             partAttributeChanged(newValue);
-        else if (document().settings().ariaReflectionForElementReferencesEnabled() && (isElementReflectionAttribute(name) || isElementsArrayReflectionAttribute(name))) {
```
Why is it OK to drop this check? The flag still exists and can be toggled from the experimental features menu AFAIK.

nt1m replied:
The popover attribute uses a different flag, I could either add both flags to the check, or remove the check entirely.

cdumez replied:
You cannot remove the flag check for an aria feature like that, without removing the feature flag entirely (which you didn't). And I think it may be too early to drop the aria feature flag since the feature is super recent.

pxlcoder commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+{
+    // Store indices to avoid having to do O(n) search for every popover invoker.
+    HashMap<Element*, size_t> topLayerPositions;
```
While this instance appears safe, I think we discourage the use of raw pointers in this manner.

cc @rniwa

rniwa replied:
Yeah, definitely don't use raw pointers. Use RefPtr<Element> or Ref<Element>.

cdumez commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -321,6 +322,80 @@ String HTMLFormControlElement::resultForDialogSubmit() const
+{
+    auto value = attributeWithoutSynchronization(HTMLNames::popovertargetactionAttr);
+
+    if (equalIgnoringASCIICase(value, showAtom()))
```
Does this really need to be case insensitive?

nt1m replied:
It is case insensitive in Chrome, popovertargetaction="SHow" seems to work there.

annevk replied:
Enumerated attribute values are always case-insensitive.

cdumez commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -321,6 +322,80 @@ String HTMLFormControlElement::resultForDialogSubmit() const
+
+    if (equalIgnoringASCIICase(value, showAtom()))
+        return showAtom();
+    if (equalIgnoringASCIICase(value, hideAtom()))
```
ditto.

nt1m replied:
ditto

cdumez commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -321,6 +322,80 @@ String HTMLFormControlElement::resultForDialogSubmit() const
+// https://html.spec.whatwg.org/#popover-target-attribute-activation-behavior
+void HTMLFormControlElement::handlePopoverTargetAction() const
+{
+    auto* target = popoverTargetElement();
```
Please store in a RefPtr.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8992,15 +8993,41 @@ void Document::handlePopoverLightDismiss(PointerEvent& event)
-        return nullptr;
+        auto [clickedPopover, invokerPopover] = [&]() {
+            HTMLElement* clickedPopover = nullptr;
+            HTMLElement* invokerPopover = nullptr;
```
Since this is not perf critical code paths, please use RefPrr instead.

rniwa commented on Source/WebCore/html/HTMLFormControlElement.h:
```diff
@@ -95,6 +95,11 @@ class HTMLFormControlElement : public HTMLElement, public ValidatedFormListedEle
 
+    HTMLElement* popoverTargetElement() const;
+    const AtomString& popoverTargetAction() const;
+    void setPopoverTargetAction(const AtomString& value) { setAttributeWithoutSynchronization(HTMLNames::popovertargetactionAttr, value); }
```
It doesn't seem like it's necessary to inline this?

rniwa commented on Source/WebCore/html/PopoverInvokerElement.idl:
```diff
@@ -0,0 +1,30 @@
+[EnabledBySetting=PopoverAttributeEnabled]
+interface mixin PopoverInvokerElement {
+    [CEReactions, Reflect=popovertarget] attribute Element? popoverTargetElement;
+    [CEReactions] attribute [AtomString] DOMString popoverTargetAction;
```
It seems that we're missing CEReactions tests. We should add those at some point.

nt1m replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=252218 for the CE reaction tests, I'll cover the invoker attributes but also the popover attribute itself

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8992,15 +8993,41 @@ void Document::handlePopoverLightDismiss(PointerEvent& event)
+        auto [clickedPopover, invokerPopover] = [&]() {
+            HTMLElement* clickedPopover = nullptr;
+            HTMLElement* invokerPopover = nullptr;
+            auto isShowingPopover = [](HTMLElement& element) -> bool {
```
Note to self: rename to `isShowingAutoPopover` since this intentionally excludes manual popovers.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+        topLayerPositions.add(element, i);
+        i++;
     }
 
```
The spec says:
> [Set](https://infra.spec.whatwg.org/#map-set) popoverPositions[newPopover] to index.

Why isn't it needed for us?

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
     }
 
-    // FIXME: Include popover invokers and compare them in top layer order once they are implemented.
+    auto nearestInclusiveOpenPopover = [](Element& candidate) -> HTMLElement* {
```
Would be good to add a link to https://html.spec.whatwg.org/#nearest-inclusive-open-popover given that this is what this is implementing.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
-    return nullptr;
+    RefPtr topmostAncestor = nearestInclusiveOpenPopover(newPopover);
+
+    // Iterate over all popover invokers in the document.
```
The spec says that popover invokers are buttons. However, I don't see the logic in the loop below. We are iterating over all HTMLFormControlElement, which is not quite the same.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+    // Iterate over all popover invokers in the document.
+    for (auto& invoker : descendantsOfType<HTMLFormControlElement>(newPopover.document())) {
+        if (invoker.popoverTargetElement() == &newPopover) {
+            RefPtr candidatePopover = nearestInclusiveOpenPopover(invoker);
```
Can we add a checkAncestor() lambda or function to match the spec more closely?

cdumez commented on Source/WebCore/html/HTMLFormControlElement.h:
```diff
@@ -95,6 +95,11 @@ class HTMLFormControlElement : public HTMLElement, public ValidatedFormListedEle
 
     virtual String resultForDialogSubmit() const;
 
+    HTMLElement* popoverTargetElement() const;
```
Do these really all need to be public?

nt1m replied:
Yes, for the IDL generated code.

cdumez replied:
Really, how about handlePopoverTargetAction?

nt1m replied:
Fixed

cdumez reviewed (commented):
Please try to match the spec more closely to facilitate review and improve confidence that we match the specification.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,55 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
+        if (!is<HTMLElement>(element) || downcast<HTMLElement>(element.get()).popoverState() != PopoverState::Auto)
             continue;
-        return &element;
+        topLayerPositions.add(element, i);
```
nit: could use `i++` here.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1234,19 +1234,55 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
             continue;
-        return &element;
+        topLayerPositions.add(element, i);
+        i++;
```
And omit this.

nt1m committed:
[popover] Implement popovertarget & popovertargetaction attributes
https://bugs.webkit.org/show_bug.cgi?id=252078
rdar://105612966

Reviewed by Chris Dumez.

This adds support for the popoverTargetElement & popoverTargetAction IDL attributes, which map to popovertarget & popovertargetaction HTML attributes.

Spec change is at https://github.com/whatwg/html/pull/8962 , WPT have already been updated by a Chromium commit to match.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-invoking-attribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-stacking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-combinations-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
Rebaseline tests with progressions. Remaining failures are either due to these unimplemented parts:
- invoker relationship removal handling (not standardized?)
- popover focusing steps (https://bugs.webkit.org/show_bug.cgi?id=253231)
- anchor attribute (not standardized)

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
Update list of built files.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
Check for invokers as described in algorithm in: https://html.spec.whatwg.org/#topmost-clicked-popover

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::addRelations):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute):
(WebCore::Element::isElementsArrayReflectionAttribute):
(WebCore::Element::attributeChanged):
(WebCore::Element::getElementAttribute const):
(WebCore::Element::setElementAttribute):
(WebCore::Element::getElementsArrayAttribute const):
(WebCore::Element::setElementsArrayAttribute):
* Source/WebCore/dom/Element.h:
Add IDL element reflection support for popovertarget attribute. Simplified settings checks.

* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::invoker const):
(WebCore::PopoverData::setInvoker):
Store invoker in popover data.

* Source/WebCore/html/HTMLAttributeNames.in: Define popovertargetAttr & popovertargetActionAttr.
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler): Invoke popover activation behavior.
* Source/WebCore/html/HTMLButtonElement.idl: Use PopoverInvokerElement mixin.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::topmostPopoverAncestor): Check for invokers as defined in the spec: https://html.spec.whatwg.org/#topmost-popover-ancestor
(WebCore::HTMLElement::hidePopoverInternal): Clear invoker as defined in the spec.
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::toggleAtom):
(WebCore::showAtom):
(WebCore::hideAtom):
(WebCore::HTMLFormControlElement::popoverTargetElement const): Implement https://html.spec.whatwg.org/#popover-target-element
(WebCore::HTMLFormControlElement::popoverTargetAction const): Implement attribute reflection.
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const): Implement activation behavior: https://html.spec.whatwg.org/#popover-target-attribute-activation-behavior
* Source/WebCore/html/HTMLFormControlElement.h:
(WebCore::HTMLFormControlElement::setPopoverTargetAction): Implement attribute reflection.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler): Invoke popover activation behavior.
* Source/WebCore/html/HTMLInputElement.idl: Use PopoverInvokerElement mixin.
* Source/WebCore/html/PopoverInvokerElement.idl: Added.

Canonical link: https://commits.webkit.org/261346@main
